components/supervisor/pkg/terminal/service.go

```diff
@@ -168,18 +217,58 @@ func (srv *MuxTerminalService) Listen(req *api.ListenTerminalRequest, resp api.T
 		buf := make([]byte, 4096)
 		for {
 			n, err := stdout.Read(buf)
+			if err == io.EOF {
+				break
+			}
 			if err != nil {
 				errchan <- err
 				return
 			}
 
-			// TODO(cw): find out how to separate stdout/stderr
-			err = resp.Send(&api.ListenTerminalResponse{Output: &api.ListenTerminalResponse_Stdout{Stdout: buf[:n]}})
+			err = resp.Send(&api.ListenTerminalResponse{Output: &api.ListenTerminalResponse_Data{Data: buf[:n]}})
 			if err != nil {
 				errchan <- err
 				return
 			}
 		}
+
+		state, err := term.Wait()
+		if err != nil {
+			errchan <- err
+			return
+		}
+
+		err = resp.Send(&api.ListenTerminalResponse{Output: &api.ListenTerminalResponse_ExitCode{ExitCode: int32(state.ExitCode())}})
+		if err != nil {
+			errchan <- err
+			return
+		}
+		errchan <- io.EOF
+	}()
+	go func() {
+		title, _ := term.Title()
+		err := resp.Send(&api.ListenTerminalResponse{Output: &api.ListenTerminalResponse_Title{Title: title}})
+		if err != nil {
+			errchan <- err
+			return
+		}
+		for {
+			select {
+			case <-resp.Context().Done():
+				return
+			case <-time.After(200 * time.Millisecond):
```

Comment:
this will create an awful many `time.Ticker`. Better create a ticker ahead of time and use that one:
```Go
t := time.NewTicker(200 * time.Millisecond)
defer t.Stop()
for {
    select {
        case <-t.C:
            ...
    }
}
```